src/ivectorbin/ivector-plda-scoring-dense.cc

```diff
@@ -0,0 +1,226 @@
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "util/stl-utils.h"
+#include "ivector/plda.h"
+
+namespace kaldi {
+
+bool EstPca(const Matrix<BaseFloat> &ivector_mat, BaseFloat target_energy,
+  Matrix<BaseFloat> *mat) {
+  int32 num_rows = ivector_mat.NumRows(),
+    num_cols = ivector_mat.NumCols();
+  Vector<BaseFloat> sum;
+  SpMatrix<BaseFloat> sumsq;
+  sum.Resize(num_cols);
+  sumsq.Resize(num_cols);
+  sum.AddRowSumMat(1.0, ivector_mat);
+  sumsq.AddMat2(1.0, ivector_mat, kTrans, 1.0);
+  sum.Scale(1.0 / num_rows);
+  sumsq.Scale(1.0 / num_rows);
+  sumsq.AddVec2(-1.0, sum); // now sumsq is centered covariance.
+  int32 full_dim = sum.Dim();
+
+  Matrix<BaseFloat> P(full_dim, full_dim);
+  Vector<BaseFloat> s(full_dim);
+
+  try {
+    if (num_rows > num_cols)
+      sumsq.Eig(&s, &P);
+    else
+      Matrix<BaseFloat>(sumsq).Svd(&s, &P, NULL);
+  } catch (...) {
+    return false;
+  }
+
+  SortSvd(&s, &P);
+
+  Matrix<BaseFloat> transform(P, kTrans); // Transpose of P.  This is what
+                                       // appears in the transform.
+  Vector<BaseFloat> offset(full_dim);
+
+  // We want the PCA transform to retain target_energy amount of the total
+  // energy.
+  BaseFloat total_energy = s.Sum();
+  BaseFloat energy = 0.0;
+  int32 dim = 1;
+  while (energy / total_energy <= target_energy) {
+    energy += s(dim-1);
+    dim++;
+  }
+  Matrix<BaseFloat> transform_float(transform);
+  mat->Resize(transform.NumCols(), transform.NumRows());
+  mat->CopyFromMat(transform);
+  mat->Resize(dim, transform_float.NumCols(), kCopyData);
+  return true;
+}
+
+void TransformIvectors(const Matrix<BaseFloat> &ivectors_in,
+  const PldaConfig &plda_config, Plda *plda,
+  Matrix<BaseFloat> *ivectors_out) {
+  int32 dim = plda->Dim();
+  ivectors_out->Resize(ivectors_in.NumRows(), dim);
+  for (int32 i = 0; i < ivectors_in.NumRows(); i++) {
+    Vector<BaseFloat> transformed_ivector(dim);
+    plda->TransformIvector(plda_config, ivectors_in.Row(i), 1.0,
+      &transformed_ivector);
+    ivectors_out->Row(i).CopyFromVec(transformed_ivector);
+  }
+}
+
+void ApplyPca(const Matrix<BaseFloat> &ivector_mat,
+  const Matrix<BaseFloat> &pca_mat, Matrix<BaseFloat> *ivector_mat_out) {
+
+  int32 transform_cols = pca_mat.NumCols(),
+        transform_rows = pca_mat.NumRows(),
+        feat_dim = ivector_mat.NumCols();
+  ivector_mat_out->Resize(ivector_mat.NumRows(), transform_rows);
+  KALDI_ASSERT(transform_cols == feat_dim);
+  ivector_mat_out->AddMatMat(1.0, ivector_mat, kNoTrans,
+    pca_mat, kTrans, 0.0);
+}
+
+}
+
+int main(int argc, char *argv[]) {
+  using namespace kaldi;
+  typedef kaldi::int32 int32;
+  typedef kaldi::int64 int64;
+  try {
+    const char *usage =
+      "Perform PLDA scoring for speaker diarization.  The input spk2utt\n"
```

Comment:
is this really a spk2utt file?  It seems like it represents something different.  I wonder if renaming it somehow would increase clarity.

Reply:
reco2utt is the notation used in other places such as for segmentation.